service/worker/workerdeployment/workflow.go

```diff
@@ -483,21 +488,36 @@ func (d *WorkflowRunner) validateAddVersionToWorkerDeployment(args *deploymentsp
 			return temporal.NewApplicationError("deployment version already registered", errVersionAlreadyExistsType)
 		}
 	}
-
 	return nil
 }
 
+func (d *WorkflowRunner) getMaxVersions(ctx workflow.Context) int {
+	getMaxVersionsInDeployment := func(ctx workflow.Context) interface{} {
+		return d.unsafeMaxVersion()
+	}
+	intEq := func(a, b interface{}) bool {
+		return a == b
+	}
+	var maxVersions int
+	if err := workflow.MutableSideEffect(ctx, "getMaxVersions", getMaxVersionsInDeployment, intEq).Get(&maxVersions); err != nil {
+		// This should not happen really. but just in case.
+		return defaultMaxVersions
+	}
+	return maxVersions
+}
+
 func (d *WorkflowRunner) handleAddVersionToWorkerDeployment(ctx workflow.Context, args *deploymentspb.AddVersionUpdateArgs) error {
 	d.pendingUpdates++
 	defer func() {
 		d.pendingUpdates--
 	}()
-	if len(d.State.Versions) >= defaultMaxVersions {
-		err := d.tryDeleteVersion(ctx)
-		if err != nil {
-			return err
-		}
+
+	maxVersions := d.getMaxVersions(ctx)
+
+	if len(d.State.Versions) >= maxVersions {
+		return temporal.NewApplicationError(fmt.Sprintf("cannot add version, already at max versions %d", maxVersions), errTooManyVersions)
```

Comment:
```suggestion
                err := d.tryDeleteVersion(ctx)
                if err != nil {
                		return temporal.NewApplicationError(fmt.Sprintf("cannot add version, already at max versions %d", maxVersions), errTooManyVersions)
                }
```

Reply:
important

Reply:
the test that tests this was skipped.. which is why it didn't catch this regfression

Reply:
Oh, good catch!